File: app/parser.py
"""Parse property information from real estate / listing web pages."""

import json
import re
from typing import Any, Optional

from bs4 import BeautifulSoup


def _text(soup: Optional[BeautifulSoup]) -> str:
    if soup is None:
        return ""
    return soup.get_text(strip=True) or ""


def _first_group(text: str, pattern: str) -> str:
    m = re.search(pattern, text, re.I)
    return m.group(1).strip() if m else ""


def extract_json_ld(soup: BeautifulSoup) -> list[dict[str, Any]]:
    """Extract all JSON-LD script blocks (schema.org often used for listings)."""
    scripts = soup.find_all("script", type="application/ld+json")
    result = []
    for s in scripts:
        if not s.string:
            continue
        try:
            data = json.loads(s.string)
            if isinstance(data, list):
                result.extend(data)
            else:
                result.append(data)
        except json.JSONDecodeError:
            continue
    return result


def parse_properties_from_html(html: str, url: str = "") -> dict[str, Any]:
    """
    Parse property-related information from HTML.
    Handles JSON-LD (schema.org), meta tags, and common listing patterns.
    """
    soup = BeautifulSoup(html, "html.parser")
    out: dict[str, Any] = {
        "url": url,
        "address": "",
        "price": "",
        "beds": "",
        "baths": "",
        "sqft": "",
        "description": "",
        "title": "",
        "images": [],
        "raw_json_ld": [],
        "meta": {},
    }

    # --- JSON-LD (RealEstateAgent, Product, Place, etc.) ---
    for ld in extract_json_ld(soup):
        out["raw_json_ld"].append(ld)
        t = ld.get("@type", "")
        if isinstance(t, list):
            t = t[0] if t else ""
        if "Place" in str(t) or "RealEstate" in str(t) or "Product" in str(t):
            addr = ld.get("address")
            if isinstance(addr, dict):
                out["address"] = out["address"] or addr.get("streetAddress", "")
            elif addr and not out["address"]:
                out["address"] = str(addr)
            if isinstance(ld.get("address"), dict):
                addr = ld["address"]
                if not out["address"]:
                    out["address"] = addr.get("streetAddress", "")
                out["meta"]["address_region"] = addr.get("addressRegion", "")
                out["meta"]["address_locality"] = addr.get("addressLocality", "")
            out["price"] = out["price"] or str(ld.get("offers", {}).get("price", "")) if isinstance(ld.get("offers"), dict) else out["price"]
            if isinstance(ld.get("offers"), dict) and not out["price"]:
                out["price"] = str(ld["offers"].get("price", ""))
            out["description"] = out["description"] or _text(BeautifulSoup(ld.get("description", ""), "html.parser"))
            out["title"] = out["title"] or ld.get("name", "")
        if "Product" in str(t):
            offers = ld.get("offers")
            if isinstance(offers, dict) and not out["price"]:
                out["price"] = str(offers.get("price", ""))
            elif isinstance(offers, list) and offers and not out["price"]:
                out["price"] = str(offers[0].get("price", ""))

    # --- Meta tags (og:, twitter:, description) ---
    for meta in soup.find_all("meta", property=re.compile(r"^(og:|twitter:)")):
        prop = meta.get("property") or meta.get("name") or ""
        content = meta.get("content", "")
        if "title" in prop.lower():
            out["title"] = out["title"] or content
        elif "description" in prop.lower():
            out["description"] = out["description"] or content
        elif "image" in prop.lower() and content and content not in out["images"]:
            out["images"].append(content)
    desc_meta = soup.find("meta", attrs={"name": "description"})
    if desc_meta and not out["description"]:
        out["description"] = desc_meta.get("content", "")

    # --- Common listing patterns in text ---
    body_text = _text(soup.find("body"))
    if not out["price"]:
        out["price"] = _first_group(body_text, r"(\$[\d,]+(?:\.\d{2})?)")
    if not out["beds"]:
        out["beds"] = _first_group(body_text, r"(\d+)\s*(?:bed|bd|br)\b")
    if not out["baths"]:
        out["baths"] = _first_group(body_text, r"(\d+(?:\.\d+)?)\s*(?:bath|ba)\b")
    if not out["sqft"]:
        m = re.search(r"([\d,]+)\s*sq\.?\s*ft\.?|([\d,]+)\s*sf\b", body_text, re.I)
        out["sqft"] = (m.group(1) or m.group(2) or "").strip() if m else ""
    if not out["sqft"]:
        out["sqft"] = _first_group(body_text, r"(\d[\d,]*)\s*sqft")

    # --- Title from page ---
    if not out["title"]:
        t = soup.find("title")
        out["title"] = _text(t) if t else ""

    # --- Images from common listing markup ---
    for img in soup.find_all("img", src=True)[:20]:
        src = img.get("src", "")
        if src and ("photo" in src.lower() or "image" in src.lower() or "listing" in src.lower() or "property" in src.lower()):
            if src not in out["images"]:
                out["images"].append(src)
    if not out["images"]:
        for img in soup.find_all("img", src=True)[:5]:
            src = img.get("src", "")
            if src and not any(x in src for x in ["logo", "icon", "pixel", "track"]):
                out["images"].append(src)

    # Clean empty lists for nicer JSON
    if not out["raw_json_ld"]:
        del out["raw_json_ld"]
    if not out["meta"]:
        del out["meta"]
    return out

File: src/parser.rs
use regex::Regex;
use scraper::{Html, Selector};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct PropertyData {
    pub url: String,
    #[serde(skip_serializing_if = "String::is_empty")]
    pub address: String,
    #[serde(skip_serializing_if = "String::is_empty")]
    pub price: String,
    #[serde(skip_serializing_if = "String::is_empty")]
    pub beds: String,
    #[serde(skip_serializing_if = "String::is_empty")]
    pub baths: String,
    #[serde(skip_serializing_if = "String::is_empty")]
    pub sqft: String,
    #[serde(skip_serializing_if = "String::is_empty")]
    pub description: String,
    #[serde(skip_serializing_if = "String::is_empty")]
    pub title: String,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub images: Vec<String>,
    #[serde(skip_serializing_if = "HashMap::is_empty")]
    pub meta: HashMap<String, String>,
}

fn text_from_element(doc: &Html, selector: &Selector) -> String {
    doc.select(selector)
        .next()
        .map(|e| e.text().collect::<String>().trim().to_string())
        .unwrap_or_default()
}

fn first_group(text: &str, pattern: &str) -> String {
    Regex::new(pattern)
        .ok()
        .and_then(|re| re.captures(text))
        .and_then(|caps| caps.get(1))
        .map(|m| m.as_str().trim().to_string())
        .unwrap_or_default()
}

fn extract_json_ld(doc: &Html) -> Vec<serde_json::Value> {
    let script_selector = Selector::parse("script[type='application/ld+json']").unwrap();
    let mut result = Vec::new();

    for script in doc.select(&script_selector) {
        if let Some(text) = script.text().next() {
            if let Ok(json) = serde_json::from_str::<serde_json::Value>(text) {
                match json {
                    serde_json::Value::Array(arr) => result.extend(arr),
                    val => result.push(val),
                }
            }
        }
    }
    result
}

pub fn parse_properties_from_html(html: &str, url: &str) -> PropertyData {
    let doc = Html::parse_document(html);
    let mut out = PropertyData {
        url: url.to_string(),
        address: String::new(),
        price: String::new(),
        beds: String::new(),
        baths: String::new(),
        sqft: String::new(),
        description: String::new(),
        title: String::new(),
        images: Vec::new(),
        meta: HashMap::new(),
    };

    // Extract JSON-LD
    for ld in extract_json_ld(&doc) {
        if let Some(type_val) = ld.get("@type") {
            let type_str = match type_val {
                serde_json::Value::String(s) => s.clone(),
                serde_json::Value::Array(arr) => {
                    arr.first().and_then(|v| v.as_str()).unwrap_or("").to_string()
                }
                _ => String::new(),
            };

            if type_str.contains("Place") || type_str.contains("RealEstate") || type_str.contains("Product") {
                // Address
                if let Some(addr) = ld.get("address") {
                    if let Some(street) = addr.get("streetAddress").and_then(|v| v.as_str()) {
                        if out.address.is_empty() {
                            out.address = street.to_string();
                        }
                    }
                    if let Some(region) = addr.get("addressRegion").and_then(|v| v.as_str()) {
                        out.meta.insert("address_region".to_string(), region.to_string());
                    }
                    if let Some(locality) = addr.get("addressLocality").and_then(|v| v.as_str()) {
                        out.meta.insert("address_locality".to_string(), locality.to_string());
                    }
                }

                // Price from offers
                if let Some(offers) = ld.get("offers") {
                    if let Some(price_val) = offers.get("price") {
                        if out.price.is_empty() {
                            out.price = price_val.to_string().trim_matches('"').to_string();
                        }
                    }
                }

                // Description
                if let Some(desc) = ld.get("description").and_then(|v| v.as_str()) {
                    if out.description.is_empty() {
                        out.description = desc.to_string();
                    }
                }

                // Title/Name
                if let Some(name) = ld.get("name").and_then(|v| v.as_str()) {
                    if out.title.is_empty() {
                        out.title = name.to_string();
                    }
                }
            }
        }
    }

    // Meta tags (og:, twitter:, description)
    let meta_selector = Selector::parse("meta").unwrap();
    for meta in doc.select(&meta_selector) {
        let prop = meta
            .value()
            .attr("property")
            .or_else(|| meta.value().attr("name"))
            .unwrap_or("");
        let content = meta.value().attr("content").unwrap_or("");

        if prop.to_lowercase().contains("title") && out.title.is_empty() {
            out.title = content.to_string();
        } else if prop.to_lowercase().contains("description") && out.description.is_empty() {
            out.description = content.to_string();
        } else if prop.to_lowercase().contains("image") && !content.is_empty() && !out.images.contains(&content.to_string()) {
            out.images.push(content.to_string());
        }
    }

    // Title from page
    if out.title.is_empty() {
        let title_selector = Selector::parse("title").unwrap();
        out.title = text_from_element(&doc, &title_selector);
    }

    // Body text for pattern matching
    let body_selector = Selector::parse("body").unwrap();
    let body_text = text_from_element(&doc, &body_selector);

    // Extract price
    if out.price.is_empty() {
        out.price = first_group(&body_text, r"(\$[\d,]+(?:\.\d{2})?)");
    }

    // Extract beds
    if out.beds.is_empty() {
        out.beds = first_group(&body_text, r"(\d+)\s*(?:bed|bd|br)\b");
    }

    // Extract baths
    if out.baths.is_empty() {
        out.baths = first_group(&body_text, r"(\d+(?:\.\d+)?)\s*(?:bath|ba)\b");
    }

    // Extract sqft
    if out.sqft.is_empty() {
        let sqft_re = Regex::new(r"([\d,]+)\s*sq\.?\s*ft\.?|([\d,]+)\s*sf\b").unwrap();
        if let Some(caps) = sqft_re.captures(&body_text) {
            out.sqft = caps
                .get(1)
                .or_else(|| caps.get(2))
                .map(|m| m.as_str().trim().to_string())
                .unwrap_or_default();
        }
    }
    if out.sqft.is_empty() {
        out.sqft = first_group(&body_text, r"(\d[\d,]*)\s*sqft");
    }

    // Extract images
    let img_selector = Selector::parse("img[src]").unwrap();
    for img in doc.select(&img_selector).take(20) {
        if let Some(src) = img.value().attr("src") {
            let src_lower = src.to_lowercase();
            if src_lower.contains("photo")
                || src_lower.contains("image")
                || src_lower.contains("listing")
                || src_lower.contains("property")
            {
                if !out.images.contains(&src.to_string()) {
                    out.images.push(src.to_string());
                }
            }
        }
    }

    if out.images.is_empty() {
        for img in doc.select(&img_selector).take(5) {
            if let Some(src) = img.value().attr("src") {
                let src_lower = src.to_lowercase();
                if !src_lower.contains("logo")
                    && !src_lower.contains("icon")
                    && !src_lower.contains("pixel")
                    && !src_lower.contains("track")
                {
                    out.images.push(src.to_string());
                }
            }
        }
    }

    out
}
